Split test and development setup

yvanzo opened the pull request:
This patch distinguishes _test_ setup from _development_ setup:
    
* test setup to run server in standalone mode with sample data,
* development setup to run any local working copy of musicbrainz-server.

yvanzo committed:
Rename 'development' setup to 'test' setup

It allows to distinguish between test setup and development setup:

* test setup to run server in standalone mode with sample data,
  purposed for application developers;
* development setup to run local working copy of musicbrainz-server,
  purposed for MusicBrainz Server developers/contributors.

yvanzo committed:
Add build/compose files for local development

It is purposed for MusicBrainz Server developers and contributors.

It allows running local development version of MusicBrainz Server by
mounting local working copy from `MUSICBRAINZ_SERVER_LOCAL_ROOT` onto
`/musicbrainz-server` in the container for the `musicbrainz` service.

See https://docs.docker.com/compose/faq/#should-i-include-my-code-with-copyadd-or-a-volume

New `build/musicbrainz-dev/` heavily derives from `build/musicbrainz/`.
Scripts are almost the same but `start.sh` which installs/updates
JavaScript and Perl dependencies in `musicbrainz-server/node_modules`
and `musicbrainz-server/perl_modules` respectively. This custom path
Perl modules are installed in must be given when setting environment:

    eval $( perl -Mlocal::lib="${MUSICBRAINZ_PERL_LOCAL_LIB}" )

yvanzo committed:
Set PROD_STANDY to read the production database

This is a database definition for use by official maintainers only.

Note that standby/slave/secondary database is read-only anyway, thus
setting user to `musicbrainz` or musicbrainz_ro` makes no difference.
It is set to `musicbrainz_ro` as a precaution just in case the SSH
tunnel mistakenly points to a database other than the standby prod.

It relies on https://github.com/metabrainz/musicbrainz-server/pull/1337

yvanzo committed:
Run a local copy of the HTML validator

This is used when running tests that use `html_ok`, for example:

    prove -l t/tests.t :: --tests 'Controller::Work::Tags'

yvanzo committed:
Install language packs

This is needed to deploy localization PO files under po/ directory.

yvanzo committed:
Enable local Perl modules for interactive bash

yvanzo committed:
Allow setting dev flags from Docker environment

* DBDefs->CATALYST_DEBUG now defaults to 0 and can be set through
  Docker MUSICBRAINZ_CATALYST_DEBUG environment variable.
  (Setting it to 1 currently seems to bug mini-relationships editor.)

* DBDefs->DEVELOPMENT_SERVER still defaults to 1 and can be set through
  Docker MUSICBRAINZ_DEVELOPMENT_SERVER environment variable.

SuperSandro2000 commented on build/musicbrainz-dev/scripts/recreatedb.sh:
```diff
@@ -0,0 +1,4 @@
+#!/bin/bash
+eval $( perl -Mlocal::lib="${MUSICBRAINZ_PERL_LOCAL_LIB}" )
+
+psql postgres -U musicbrainz -h db -c "DROP DATABASE musicbrainz_db;"; /createdb.sh "$@"
```
I think that we shouldn't hardcode the credentials here.

yvanzo replied:
Agreed but this is an issue of source `build/musicbrainz/scripts/recreatedb.sh`

SuperSandro2000 commented on build/musicbrainz-dev/scripts/createdb.sh:
```diff
@@ -0,0 +1,100 @@
+#!/bin/bash
+
+eval $( perl -Mlocal::lib="${MUSICBRAINZ_PERL_LOCAL_LIB}" )
+
```
Did you miss ` set -e -u` or leave it out on purpose?

yvanzo replied:
No, it just derives from `build/musicbrainz/scripts/createdb.sh` which don’t set it either.

Note: `set -u` would potentially break here, see https://github.com/metabrainz/musicbrainz-docker/pull/123#discussion_r390221359

SuperSandro2000 replied:
Good and sad to know that.

yvanzo replied:
Rebased and followed https://github.com/metabrainz/musicbrainz-docker/commit/5e174b71238921ea045386e1730ffde02b5ea77a, see https://github.com/metabrainz/musicbrainz-docker/pull/118/commits/992af4e2db7c30247981372561f4d23358d133b1#diff-97d9c507ad8516ff5c9bc6e24001d36eR2-R7.

SuperSandro2000 commented on build/musicbrainz-dev/scripts/createdb.sh:
```diff
@@ -0,0 +1,100 @@
+case "$IMPORT" in
+    fullexport  )
+        DUMP_FILES=(
+        mbdump.tar.bz2
+        mbdump-cdstubs.tar.bz2
+        mbdump-cover-art-archive.tar.bz2
+        mbdump-derived.tar.bz2
+        mbdump-stats.tar.bz2
+        mbdump-wikidocs.tar.bz2
```
```suggestion
          mbdump.tar.bz2
          mbdump-cdstubs.tar.bz2
          mbdump-cover-art-archive.tar.bz2
          mbdump-derived.tar.bz2
          mbdump-stats.tar.bz2
          mbdump-wikidocs.tar.bz2
```

yvanzo replied:
Again, it has to be fixed in original script first, we don’t want these scripts to differ by coding style.

yvanzo replied:
This script is being split (working on PR) to allow fetching search indexes anyway.

yvanzo replied:
Fixed in https://github.com/metabrainz/musicbrainz-docker/pull/125/commits/5047754d5b5a7880667e007a04182012cc0a9f6a

SuperSandro2000 commented on build/musicbrainz-dev/Dockerfile:
```diff
@@ -0,0 +1,78 @@
+    POSTGRES_PASSWORD=musicbrainz
+
+VOLUME  ["/media/dbdump"]
+VOLUME  ["/musicbrainz-server"]
```
I am not a big fan of putting volumes in docker images as it force creates them and there is not way to turn them off. Putting them in the docker-compose.yml should be enough.

yvanzo replied:
Seems sensible, more details: https://boxboat.com/2017/01/23/volumes-and-dockerfiles-dont-mix/

We should probable remove `/media/dbdump` from `build/musicbrainz/Dockerfile` too.

yvanzo replied:
Reverted adding `VOLUME` for `/musicbrainz-server` in https://github.com/metabrainz/musicbrainz-docker/pull/118/commits/771c75edd69039ae7ed3516d0a1a64d310c987b0 (pending rebase).

yvanzo replied:
Removed `/media/dbdump` from Dockerfile in 33dc6b3393392fbcfd8a435ae66c6d22d87a58f3 on `mbvm-38-dev` branch.

yvanzo replied:
Rebased and removed both `VOLUME` instructions.

SuperSandro2000 commented on build/musicbrainz-dev/Dockerfile:
```diff
@@ -0,0 +1,78 @@
+        wget \
+        yarn \
+        zlib1g-dev && \
+    apt-get clean && \
```
The Debian and Ubuntu Images from the Docker Library run this automatically.

yvanzo replied:
This default behavior should be checked for the current base image which is neither of these.

SuperSandro2000 replied:
The metabrainz base image is based on [phusion/baseimage:0.9.22](https://github.com/phusion/baseimage-docker/blob/0.9.22/image/Dockerfile) which is based on ubuntu 16.04.
I installed git before in the docker image and now tree and didn't run any other commands in between and it run ``apt-get clean`` automatically.

````
apt install --no-install-recommends git tree
Reading package lists... Done
Building dependency tree       
Reading state information... Done
git is already the newest version (1:2.7.4-0ubuntu1.7).
The following NEW packages will be installed:
  tree
0 upgraded, 1 newly installed, 0 to remove and 104 not upgraded.
Need to get 40.6 kB of archives.
After this operation, 138 kB of additional disk space will be used.
Do you want to continue? [Y/n] 
Get:1 http://archive.ubuntu.com/ubuntu xenial/universe amd64 tree amd64 1.7.0-3 [40.6 kB]
Fetched 40.6 kB in 0s (255 kB/s)
Selecting previously unselected package tree.
(Reading database ... 11590 files and directories currently installed.)
Preparing to unpack .../tree_1.7.0-3_amd64.deb ...
Unpacking tree (1.7.0-3) ...
Setting up tree (1.7.0-3) ...
root@c1fb1a52fd02:/# tree /var/cache/apt/archives/
/var/cache/apt/archives/
├── lock
└── partial

1 directory, 1 file
root@c1fb1a52fd02:/# 
`````

Under my desktop the cache is almost 800MB:

````
sudo diskus /var/cache/apt/archives/
715.87 MB (715,874,304 bytes)
````

So it is proven that ``apt-get clean`` is run automatically.

yvanzo replied:
Removed it from while rebasing, see https://github.com/metabrainz/musicbrainz-docker/pull/118/commits/992af4e2db7c30247981372561f4d23358d133b1#diff-bfa80dfe6fc07e10f8447a73a919129cR48.

SuperSandro2000 commented on build/musicbrainz-dev/Dockerfile:
```diff
@@ -0,0 +1,78 @@
+FROM metabrainz/base-image
+LABEL Maintainer="yvanzo@metabrainz.org"
+
+ENV DOCKERIZE_VERSION v0.5.0
```
```suggestion
ENV DOCKERIZE_VERSION=v0.5.0
```
for consistency :)

yvanzo replied:
Consistency with?

SuperSandro2000 replied:
with line 64 to 70 of this file.

yvanzo replied:
That is consistent with `build/musicbrainz/Dockerfile`. We don’t want these to differ by coding style.

yvanzo commented:
Rebased on latest `mbvm-38-dev` branch. Dropped commit https://github.com/metabrainz/musicbrainz-docker/pull/118/commits/594338e1408124d491b810227cafb1851cadf52b and marked all conversations nonspecific to `musicbrainz-dev/` as resolved. These should be addressed in a further pull request.

reosarevok commented on build/musicbrainz-dev/Dockerfile:
```diff
@@ -26,6 +26,16 @@ RUN curl -sSL https://www.postgresql.org/media/keys/ACCC4CF8.asc | apt-key add -
         gettext \
         g++ \
         git \
+        language-pack-de \
+        language-pack-el \
+        language-pack-es \
+        language-pack-et \
+        language-pack-fi \
+        language-pack-fr \
+        language-pack-it \
+        language-pack-ja \
+        language-pack-nl \
+        language-pack-sq \
```
Is it useful to install so many of them? Just having a few of the most complete ones is probably enough for testing.

yvanzo replied:
Yes, these are the few most complete ones available in MusicBrainz [`tests`](https://github.com/metabrainz/musicbrainz-server/blob/v-2020-03-16/docker/Dockerfile.tests#L35-L44) image for CI, and in [`website`](https://github.com/metabrainz/musicbrainz-server/blob/v-2020-03-16/docker/templates/macros.m4#L128) image for [test](https://test.musicbrainz.org) and [beta](https://beta.musicbrainz.org) (but `sq`) websites. It can be useful to maintainer for debugging an error spotted from any of these, or just to developer for testing with favorite language.

IIRC, installing these language packs must be done ahead of installing some Node or Perl modules, that is the main reason to install these packs here rather than carry developer with the burden of maintaining a custom `Dockerfile` and reinstalling Node and Perl modules if needed.

Ideally, languages should be selectable through Docker environment, so you can just install the ones you want to work with. Reported as MBVM-50.

reosarevok reviewed (approved):
This seems generally sane, although I don't have the disc space to test it right now.
